internal/protoc/downloader.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package protoc
+
+import (
+	"archive/zip"
+	"bytes"
+	"crypto/sha512"
+	"encoding/base64"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+	"sync"
+
+	"github.com/uber/prototool/internal/x/settings"
+	"github.com/uber/prototool/internal/x/vars"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+type downloader struct {
+	logger    *zap.Logger
+	cachePath string
+	protocURL string
+	config    settings.Config
+
+	lock sync.RWMutex
+	// the looked-up and verified to exist base path
+	cachedBasePath string
+}
+
+func newDownloader(config settings.Config, options ...DownloaderOption) *downloader {
+	downloader := &downloader{
+		config: config,
+		logger: zap.NewNop(),
+	}
+	for _, option := range options {
+		option(downloader)
+	}
+	if downloader.config.Compile.ProtobufVersion == "" {
```

Comment:
Not actionable here, but we should consider setting this default before it's passed into `newDownloader`. I realized that the `settings` package doesn't concern itself with a lot of the defaults just yet, but it's something to consider moving forward.

Reply:
I don't personally like relying on structs being well-formed more than you have to assume, it's just asking for bugs down the road, so I like the default value being empty. Just my two cents.